ultra/docs/adapters.md

```diff
@@ -0,0 +1,143 @@
+        [
+            relative_vehicle_position_x / 100.0,
+            relative_vehicle_position_y / 100.0,
+            heading_difference / 3.14,
+            social_vehicle_speed / 30.0
+        ],  # 10th closest social vehicle.
+    ]
+}
+```
+
+Where:
+- `ego_vehicle_speed` is the speed of the ego vehicle in meters per second.
+- `distance_from_center` is the lateral distance between the center of the closest waypoint's lane and the ego vehicle's position, divided by half of that lane's width.
+- `steering` is the angle of ego vehicle's front wheels in radians.
+- `angle_error` is the closest waypoint's heading minus the ego vehicle's heading.
+- `relative_goal_position_x` is the x component of the vector obtained by calculating the goal's `(x, y)` position minus the ego vehicle's `(x, y)` position, and rotating that difference by the negative of the ego vehicle's heading. All in all, this keeps this component completely relative from the ego vehicle's perspective.
+- `relative_goal_position_y` is the y component of the vector obtained by calculating the goal's `(x, y)` position minus the ego vehicle's `(x, y)` position, and rotating that difference by the negative of the ego vehicle's heading. All in all, this keeps this component completely relative from the ego vehicle's perspective.
+- `relative_waypoint_position_x` is the x component of the vector obtained by calculating the waypoint's `(x, y)` position minus the ego vehicle's `(x, y)` position, and rotating that difference by the negative of the ego vehicle's heading. All in all, this keeps the component completely relative from the ego vehicle's perspective.
+- `relative_waypoint_position_y` is the y component of the vector obtained by calculating the waypoint's `(x, y)` position minus the ego vehicle's `(x, y)` position, and rotating that difference by the negative of the ego vehicle's heading. All in all, this keeps the component completely relative from the ego vehicle's perspective.
+- `road_speed` is the speed limit of the closest waypoint.
+- `relative_vehicle_position_x` is the x component of vector obtained by calculating the social vehicle's `(x, y)` position minus the ego vehicle's `(x, y)` position, and rotating that difference by the negative of the ego vehicle's heading. All in all, this keeps this component completely relative from the ego vehicle's perspective.
+- `relative_vehicle_position_y` is the y component of vector obtained by calculating the social vehicle's `(x, y)` position minus the ego vehicle's `(x, y)` position, and rotating that difference by the negative of the ego vehicle's heading. All in all, this keeps this component completely relative from the ego vehicle's perspective.
+- `heading_difference` is the heading of the social vehicle minus the heading of the ego vehicle.
+- `social_vehicle_speed` is the speed of the social vehicle in meters per second.
+
+Notice that the social vehicles are sorted by relative distance to the ego vehicle.
```

Comment:
I know we discussed this before, but we should describe why this is necessary (I think its because precog, which is a default). Regarding precog, we should also look into whether it makes sense experimentally. I skimmed the paper and it seems like some of their design decisions (like masking / padding with zeros) is specific to the image observations they use.

Reply:
Yeah it seems like with the original adaptation that ULTRA had, if an encoder was used (no matter if it was the Precog or PointNet), then the social vehicles were sorted by relative distance to the ego vehicle. Will ask Nhat why this method was chosen for the encoders.

Also yeah, good idea. Will also take another look at the Precog paper.